lib/ast.js

```diff
@@ -1,66 +1,81 @@
-        if (vulnerableFunctionNames.includes(name)) {
-          declaredFunctions[name] = loc;
-        }
-      } else if ((node.type === 'ExpressionStatement' &&
-                  node.expression.right &&
-                  node.expression.right.type === 'FunctionExpression')) {
-        loc = node.expression.right.body.loc.start;
-        // Todo: going recursively over object
-        if (node.expression.left.object.object &&
-            node.expression.left.object.object.name) {
-          name = node.expression.left.object.object.name;
-          name = name + '.' + node.expression.left.object.property.name;
-        } else {
-          name = node.expression.left.object.name;
-        }
-        name = name + '.' + node.expression.left.property.name;
-        if (vulnerableFunctionNames.includes(name)) {
-          declaredFunctions[name] = loc;
-        }
-      } else if (node.type === 'VariableDeclaration') {
-        node.declarations.forEach((decl) => {
-          if (decl.init && decl.init.type === 'FunctionExpression') {
-            if (decl.init.id && decl.init.id.name) {
-              name = decl.init.id.name;
-            } else {
-              name = decl.id.name;
-            }
-            loc = decl.init.body.loc.start;
-            if (vulnerableFunctionNames.includes(name)) {
-              declaredFunctions[name] = loc;
-            }
-          }
-        });
-      } else if (node.type === 'ExportNamedDeclaration') {
-        if (node.declaration.type === 'FunctionDeclaration') {
-          name = node.declaration.id.name;
-          loc = node.declaration.body.loc.start;
-          if (vulnerableFunctionNames.includes(name)) {
-            declaredFunctions[name] = loc;
-          }
+  const parsedScript = acorn.parse(scriptContent,
+    {locations: true, sourceType: 'module'});
+  const body = parsedScript.body;
+  body.forEach(function (node) {
+    inspectNode(node,
+      [],
+      (name, loc) => {
+        const mangled = name.join('.');
+        if (vulnerableFunctionNames.includes(mangled)) {
+          declaredFunctions[mangled] = loc;
         }
+      });
+  });
+
+  return declaredFunctions;
+}
+
+function inspectNode(node, path, cb) {
+  if (!node) {
+    return;
+  }
+  if (node.type === 'FunctionDeclaration') {
+    const loc = node.body.loc.start;
+    const name = node.id.name;
+    cb(path.concat(name), loc);
+  } else if (node.type === 'FunctionExpression') {
+    cb(path, node.body.loc.start);
+  } else if ((node.type === 'ExpressionStatement' && node.expression.right)) {
+    const name = [];
+    const left = node.expression.left;
+
+    if (!left.object) {
+      return;
+    }
+
+    if (left.object.object && left.object.object.name) {
+      name.push(left.object.object.name);
+      if (left.object.property && left.object.property.name) {
+        name.push(left.object.property.name);
+      }
+    } else {
+      name.push(left.object.name);
+    }
+    name.push(left.property.name);
```

Comment:
Not critical to this PR, because it's not your code: In most of the places in this file we protect from non existing keys, but there are places like here that we don't. So I think we should have a single convention and check that ```left.property && left.property.name``` before (and in all the other missing places)

Reply:
These were unchecked before this PR, so I left them be.

It might be more readable to filter out the `undefined`'s from the path when we turn it back into a string. This code is already ugly enough here. Maybe `_.get()` can save us. I am keen to leave it as is for this PR.